cmd/release-controller/upgrades.go

```diff
@@ -317,34 +322,173 @@ func syncGraphToSecret(graph *UpgradeGraph, update bool, secretClient kv1core.Se
+			return a.From < b.From
+		}
+		return a.To < b.To
+	})
+	for _, record := range records {
+		sort.Slice(record.Results, func(i, j int) bool {
+			return record.Results[i].URL < record.Results[j].URL
+		})
+	}
+	return records
+}
 
-	// keep the secret up to date
-	buf := &bytes.Buffer{}
-	wait.Until(func() {
-		buf.Reset()
-		if err := graph.Save(buf); err != nil {
-			klog.Errorf("Unable to calculate graph state: %v", err)
-			return
+const (
+	PruneGraphPrintSecret = "secret"
+	PruneGraphPrintDebug  = "debug"
+)
+
+func (c *Controller) pruneGraph(secretClient kv1core.SecretInterface, ns, name string, printOption string, confirm bool) error {
+	stopCh := wait.NeverStop
+
+	loadUpgradeGraph(c.graph, secretClient, ns, name, stopCh)
+
+	imageStreams, err := c.releaseLister.ImageStreams(ns).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var stableTagList []string
+
+	for _, imageStream := range imageStreams {
+		r, ok, err := c.releaseDefinition(imageStream)
+		if err != nil || !ok {
+			continue
 		}
-		secret, err := secretClient.Get(context.TODO(), name, metav1.GetOptions{})
-		if err != nil {
-			klog.Errorf("Can't read latest secret %s/%s: %v", ns, name, err)
-			return
+		if r.Config.As == releaseConfigModeStable {
+			for _, tag := range imageStream.Spec.Tags {
+				stableTagList = append(stableTagList, tag.Name)
+			}
+		}
+	}
+
+	// To tags that are not present in the stable tags
+	toMissingList := make([]string, 0, len(c.graph.to))
+	for tag := range c.graph.to {
+		if !stringInSlice(tag, stableTagList) {
+			toMissingList = append(toMissingList, tag)
+		}
+	}
+
+	// From tags that are not present in the stable tags
+	fromMissingList := make([]string, 0, len(c.graph.from))
+	for tag := range c.graph.from {
+		if !stringInSlice(tag, stableTagList) {
+			fromMissingList = append(fromMissingList, tag)
 		}
-		if secret.Data == nil {
-			secret.Data = make(map[string][]byte)
+	}
+
+	// To tags that are not in the From tags
+	pruneTagList := make([]string, 0, len(c.graph.to))
+	for _, tag := range toMissingList {
+		if !stringInSlice(tag, fromMissingList) {
+			pruneTagList = append(pruneTagList, tag)
 		}
-		secret.Data["latest"] = buf.Bytes()
-		if _, err := secretClient.Update(context.TODO(), secret, metav1.UpdateOptions{}); err != nil {
-			klog.Errorf("Can't save state to secret %s/%s: %v", ns, name, err)
+	}
+
+	// From tags that are not in the To tags
+	for _, tag := range fromMissingList {
+		if !stringInSlice(tag, toMissingList) {
+			pruneTagList = append(pruneTagList, tag)
 		}
-		klog.V(2).Infof("Saved upgrade graph state to %s/%s", ns, name)
-	}, 5*time.Minute, stopCh)
+	}
+
+	// Tags that are in both the From and To lists
+	for _, tag := range toMissingList {
+		if stringInSlice(tag, fromMissingList) {
+			pruneTagList = append(pruneTagList, tag)
+		}
+	}
+
+	klog.V(2).Infof("Pruning %d/%d tags from release controller graph\n", len(pruneTagList), len(c.graph.to))
+
+	// Prune graph
+	c.graph.lock.Lock()
```

Comment:
If you're going to do this, do it with a method that has a defer.  If removeWithLock panics you are broken

Reply:
Done